app/services/concerns/xccdf/rule_group_rules.rb

```diff
@@ -0,0 +1,112 @@
+  module RuleGroupRules
+    extend ActiveSupport::Concern
+
+    included do
+      def save_rule_group_rules
+        @op_rules_and_rule_groups = @op_rule_groups + @op_rules
+        @rule_group_rules = rgr_with_requires + rgr_with_conflicts
+
+        ::RuleGroupRule.import!(
+          @rule_group_rules.select(&:new_record?), ignore: true
+        )
+
+        rules_base = ::RuleGroupRule.joins(included_rule: :benchmark)
+        rule_groups_base = ::RuleGroupRule.joins(included_rule_group: :benchmark)
+
+        rgr_links_to_remove(rules_base).delete_all
+        rgr_links_to_remove(rule_groups_base).delete_all
+      end
+
+      def save_rule_parents
+        @rule_group_rules = rules_with_rule_group_parent
+        ::RuleGroupRule.import!(@rule_group_rules.select(&:new_record?),
+                                on_duplicate_key_update: {
+                                  conflict_target: %i[rule_id],
+                                  columns: %i[rule_group_id]
+                                })
+
+        rule_parents_base = ::RuleGroupRule.joins(rule: :benchmark)
+
+        rule_parent_links_to_remove(rule_parents_base).delete_all
+      end
+
+      private
+
+      def rgr_with_requires
+        @rgr_with_requires ||= @op_rules_and_rule_groups.flat_map do |op_r_or_rg|
+          including = rule_or_rule_group_for(ref_id: op_r_or_rg.id)
+          op_r_or_rg.requires.map do |required_ref_id|
+            requiring = rule_or_rule_group_for(ref_id: required_ref_id)
+            next unless requiring
+
+            ::RuleGroupRule.find_or_initialize_by(
+              including: including, requiring: requiring
+            )
+          end
+        end.compact
+      end
+
+      def rgr_with_conflicts
+        @rgr_with_conflicts ||= @op_rules_and_rule_groups.flat_map do |op_r_or_rg|
+          including = rule_or_rule_group_for(ref_id: op_r_or_rg.id)
+          op_r_or_rg.conflicts.map do |conflicting_ref_id|
+            conflicting = rule_or_rule_group_for(ref_id: conflicting_ref_id)
+            next unless conflicting
+
+            ::RuleGroupRule.find_or_initialize_by(
+              including: including, conflicting: conflicting
+            )
+          end
+        end.compact
+      end
+
+      def rules_with_rule_group_parent
+        @rules_with_rule_group_parent ||= @op_rules.flat_map do |op_rule|
+          rule_group = rule_group_for(ref_id: op_rule.parent_id)
+          next unless rule_group
+
+          rule = rule_for(ref_id: op_rule.id)
+          ::RuleGroupRule.find_or_initialize_by(
+            rule_group: rule_group, rule: rule
+          )
+        end.compact
+      end
+
+      def rule_parent_links_to_remove(base)
+        grouped_by_rgr = @rule_group_rules.group_by(&:rule_group_id)
+        grouped_by_rgr.reduce(ProfileRule.none) do |query, (rule_group_id, prs)|
+          query.or(
+            base.where(rule_group_id: rule_group_id)
+                .where.not(rule_id: prs.map(&:rule_id))
```

Comment:
Food for thought (possibly for future enhancement): this could likely be reworked fully into SQL, avoiding to gather ids and put them as a params. If it's possible, it should be faster. Note that there is a limit of number of params within a query, limited by the postgresql lib. But I doubt that it would be a case here.

Reply:
Yeah, rule groups are being used minimally.